### ISSUE TITLE
How to avoid duplicated export name

### ISSUE THREAD

**Issue by Jamesits** [AOSC-Archive/auch#1]
E.g. I installed two vims in parallel, and run `vim` in bash, which version should be called? How about an automated call from existing software?

**Comment by m13253** [comment]
It is managed with preference control. You can control which variant you prefer. And it works with the mechanism of package matching.

**Comment by Jamesits** [comment]
Thus we need an mechanism to set "open this file with ... by default" and a way to call other version (e.g. vimx.y) which should be set automatically by package manager. (Also consider two different package who provide the same program and they are of the same version number.)

**Comment by Jamesits** [comment]
Well we can use namespaces! 

```
#!/bin/bash
import vim7.2
vim
from vim7.1 import vim as vim7.1
vim7.1
```

**Comment by m13253** [comment]
> Well we can use namespaces!

Yes we actually have this. But it will be called `aoch env vim:7.2`. It will start a shell containing vim:7.2's filesystem. If you need to execute something instead of just bring a shell, type `aoch env vim:7.2 -- do_something.sh`.

**Comment by misaka4e21** [comment]
Will there be a way to execute the rest of a script after switching version?
For example:

```
#!/bin/bash
do_something.sh --step1
eval $(auch env vim:7.2)
do_something.sh --step2
# ...
```

Or, we have to pass an argument, of a string or a script file, containing the commands we want to run?

**Comment by m13253** [comment]
> Will there be a way to execute the rest of a script after switching version?

`auch env` is implemented as environment variables. Normally, the `virtrootfs` daemon detects which process is accessing a certain file. If we know what we are doing and want to access some other package's virtual root, we can use `auch env` to start a `bash` (or whatever) containing the corresponding environment variables to tell `virtrootfs` that I want a certain virtual root. You can always edit the environment variable to reflect changes in-process, but `auch env` makes it easy.

**Comment by Jamesits** [comment]
> auch env is implemented as environment variables.

Now we've got:

```
#!/bin/bash
aoch env vim:7.2 curl:3.3 glibc:<5.3 libev:2.3.5 balabala
# do something
pushenv
aoch env vim:7.1
# do more thing
popenv
vim some file
```

lol

**Comment by m13253** [comment]
That's right.
But remember that `auch env` starts a new bash. You can just write the following commands there.

**Comment by Jamesits** [comment]
> But remember that auch env starts a new bash. You can just write the following commands there.

So how can I use different versions of a software while preserving bash local variables? Scripting will be harder.

**Comment by m13253** [comment]
Directly modify environment variables. Then flush `virtrootfs`'s cache. This is like accessing a class's private directly, so avoid using this whenever possible.

**Comment by bobcao3** [comment]
Maybe we could let virtrootfs directly change the exported version instead of changing env? I think for an fs it has this capability
